ssl/quic/quic_impl.c

```diff
@@ -4623,6 +4632,274 @@ static QUIC_CONNECTION *create_qc_from_incoming_conn(QUIC_LISTENER *ql, QUIC_CHA
     return NULL;
 }
 
+/*
+ * Token store management
+ */
+typedef struct quic_token_st {
+    CRYPTO_REF_COUNT references;
+    uint8_t *hashkey;
+    size_t hashkey_len;
+    uint8_t *token;
+    size_t token_len;
+} QUIC_TOKEN;
+
+DEFINE_LHASH_OF_EX(QUIC_TOKEN);
+
+typedef struct ssl_token_store_st {
+    LHASH_OF(QUIC_TOKEN) *cache;
+    CRYPTO_REF_COUNT references;
+    CRYPTO_MUTEX *mutex;
+} SSL_TOKEN_STORE;
+
+static uint64_t fnv1a_hash_token(uint8_t *key, size_t len)
+{
+    uint64_t hash = 0xcbf29ce484222325ULL;
+    size_t i;
+
+    for (i = 0; i < len; i++) {
+        hash ^= key[i];
+        hash *= 0x00000100000001B3ULL;
+    }
+    return hash;
+}
+
+static unsigned long quic_token_hash(const QUIC_TOKEN *item)
+{
+    return (unsigned long)fnv1a_hash_token(item->hashkey, item->hashkey_len);
+}
+
+static int quic_token_cmp(const QUIC_TOKEN *a, const QUIC_TOKEN *b)
+{
+    if (a->hashkey_len != b->hashkey_len)
+        return 1;
+    return memcmp(a->hashkey, b->hashkey, a->hashkey_len);
+}
+
+SSL_TOKEN_STORE_HANDLE *ossl_quic_new_token_store(void)
+{
+    int ok = 0;
+    SSL_TOKEN_STORE *newcache = OPENSSL_zalloc(sizeof(SSL_TOKEN_STORE));
+
+    if (newcache == NULL)
+        goto out;
+
+    newcache->cache = lh_QUIC_TOKEN_new(quic_token_hash, quic_token_cmp);
+    if (newcache->cache == NULL)
+        goto out;
+
+    newcache->mutex = ossl_crypto_mutex_new();
+#if defined(OPENSSL_THREADS)
+    if (newcache->mutex == NULL)
+        goto out;
+#endif
+
+    if (!CRYPTO_NEW_REF(&newcache->references, 1))
+        goto out;
+
+    ok = 1;
+out:
+    if (!ok) {
+        ossl_quic_free_token_store(newcache);
+        newcache = NULL;
+    }
+    return (SSL_TOKEN_STORE_HANDLE *)newcache;
+}
+
+static void free_this_token(QUIC_TOKEN *tok)
+{
+    ossl_quic_free_peer_token((QTOK *)tok);
+}
+
+void ossl_quic_free_token_store(SSL_TOKEN_STORE_HANDLE *hdl)
+{
+    int refs;
+    SSL_TOKEN_STORE *c = (SSL_TOKEN_STORE *)hdl;
```

Comment:
Why do we do this?

Reply:
because I didn't want to expose the contents of the SSL_TOKEN_STORE structure via the user api, so I typedeffed it as typedef void SSL_TOKEN_STORE_HANDLE (simmilar to what we did for rcu locks)

Reply:
This is the wrong way to achieve that goal IMO. The contents of `SSL_TOKEN_STORE` or already private to quic_impl.c because that is where the struct definition is.

All you need to do is replace the typedef of `SSL_TOKE_STORE_HANDLE` in ssl.h with this:

```` C
typedef struct ssl_token_store_st SSL_TOKEN_STORE;
````
Then the user API can use `SSL_TOKEN_STORE` type and it is fully opaque, but you do not have to do these nasty casts in your code. This is the way virtually all opaque types are handled in OpenSSL so we should be following that pattern.

Reply:
Yeah, the `typedef struct private_foo_st PUBLIC_FOO_TYPE;` pattern is SOP in OpenSSL and makes sense here.
The only place we have casts like this is for structures passed between providers and the core, where the provider may be compiled against a different version of OpenSSL with a different layout of the structure, and then we have to be careful only only use the structure on one side of the boundary, and just use it as an opaque handle to return to the other side when not the owner of the structure.